File: services/platform-api/src/handlers/quran.rs
use axum::Json;
use axum::extract::{Path, State};
use axum::http::HeaderMap;
use serde::Serialize;
use sqlx::Row;

use crate::AppState;
use crate::types::*;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SurahInfo {
    pub surah_number: i32,
    pub ayah_count: i32,
    pub name: String,
}

pub async fn list_surahs(
    State(state): State<AppState>,
    _headers: HeaderMap,
) -> Result<Json<Vec<SurahInfo>>, ApiError> {
    let rows = sqlx::query(
        "SELECT surah_number, COUNT(*) as ayah_count
         FROM canonical_ayahs
         GROUP BY surah_number
         ORDER BY surah_number",
    )
    .fetch_all(&state.pool)
    .await?;

    let surahs = rows
        .into_iter()
        .map(|r| SurahInfo {
            surah_number: r.try_get::<i32, _>("surah_number").unwrap_or(0),
            ayah_count: r
                .try_get::<i64, _>("ayah_count")
                .map(|c| c as i32)
                .unwrap_or(0),
            name: format!("Surah {}", r.try_get::<i32, _>("surah_number").unwrap_or(0)),
        })
        .collect();

    Ok(Json(surahs))
}

pub async fn get_surah(
    State(state): State<AppState>,
    _headers: HeaderMap,
    Path(surah_number): Path<i32>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let rows = sqlx::query(
        "SELECT id, surah_number, ayah_number, text_uthmani, source_checksum
         FROM canonical_ayahs WHERE surah_number = $1 ORDER BY ayah_number",
    )
    .bind(surah_number)
    .fetch_all(&state.pool)
    .await?;

    if rows.is_empty() {
        return Err(ApiError::NotFound);
    }

    let ayahs: Vec<serde_json::Value> = rows
        .into_iter()
        .map(|r| {
            serde_json::json!({
                "id": r.try_get::<String, _>("id").unwrap_or_default(),
                "surahNumber": r.try_get::<i32, _>("surah_number").unwrap_or(0),
                "ayahNumber": r.try_get::<i32, _>("ayah_number").unwrap_or(0),
                "text": r.try_get::<String, _>("text_uthmani").unwrap_or_default(),
                "sourceChecksum": r.try_get::<String, _>("source_checksum").unwrap_or_default(),
            })
        })
        .collect();

    Ok(Json(serde_json::json!({
        "surahNumber": surah_number,
        "ayahs": ayahs,
    })))
}

pub async fn get_ayah(
    State(state): State<AppState>,
    _headers: HeaderMap,
    Path((surah_number, ayah_number)): Path<(i32, i32)>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let row = sqlx::query(
        "SELECT id, surah_number, ayah_number, text_uthmani, source_checksum
         FROM canonical_ayahs
         WHERE surah_number = $1 AND ayah_number = $2",
    )
    .bind(surah_number)
    .bind(ayah_number)
    .fetch_optional(&state.pool)
    .await?
    .ok_or(ApiError::NotFound)?;

    let ayah_id: String = row.try_get("id")?;

    let words = sqlx::query(
        "SELECT id, word_index, text_uthmani, source_checksum
         FROM canonical_words WHERE ayah_id = $1 ORDER BY word_index",
    )
    .bind(&ayah_id)
    .fetch_all(&state.pool)
    .await?;

    let words_json: Vec<serde_json::Value> = words
        .into_iter()
        .map(|w| {
            serde_json::json!({
                "id": w.try_get::<String, _>("id").unwrap_or_default(),
                "wordIndex": w.try_get::<i32, _>("word_index").unwrap_or(0),
                "text": w.try_get::<String, _>("text_uthmani").unwrap_or_default(),
                "sourceChecksum": w.try_get::<String, _>("source_checksum").unwrap_or_default(),
            })
        })
        .collect();

    Ok(Json(serde_json::json!({
        "id": ayah_id,
        "surahNumber": row.try_get::<i32, _>("surah_number").unwrap_or(0),
        "ayahNumber": row.try_get::<i32, _>("ayah_number").unwrap_or(0),
        "text": row.try_get::<String, _>("text_uthmani").unwrap_or_default(),
        "sourceChecksum": row.try_get::<String, _>("source_checksum").unwrap_or_default(),
        "words": words_json,
    })))
}

File: services/node-api/routes/quran.mjs
/**
 * N9 — the canonical Quran corpus. Port of services/platform-api/src/handlers/quran.rs.
 *
 * ── No actor, and no tenant transaction. Both are deliberate ────────────────────────────────────
 * The Rust handlers take `_headers: HeaderMap` and ignore it: canonical scripture is public. And
 * they query `&state.pool` DIRECTLY rather than through `begin_tenant_tx`, because
 * `canonical_ayahs` / `canonical_surahs` / `canonical_words` are global reference tables with no
 * `tenant_id` column. Wrapping them in `withTenant` would not be "safer" — it would set a GUC no
 * policy on these tables reads, and would imply a tenant scoping that does not exist.
 *
 * ── Canonical text is never touched ─────────────────────────────────────────────────────────────
 * The text goes from the driver to the response with no `.trim()`, no `.normalize()`, no
 * `replace`. Ayah 1:1 in this corpus begins with U+FEFF and the corpus is NFC-unstable, so both of
 * those would be real, silent mutations of scripture. `tests/api-parity/quran-parity.test.mjs`
 * compares the served bytes against the row in Postgres for exactly this reason.
 */
import { NotFound, RejectionError } from "../lib/authz.mjs";

/**
 * Port of axum's `Path<i32>` extraction, INCLUDING its rejection.
 *
 * Axum parses the segment before the handler runs, so a bad value is a 400 in `text/plain` — not
 * the JSON `{"error": ...}` every other failure uses, and emphatically not a 404. A client that
 * distinguishes "no such surah" from "that is not a number" branches on exactly this.
 *
 * Matches `i32::from_str`: an optional sign, then digits, then in-range. `"1.5"`, `" "` and
 * `"99999999999999999999"` are all rejected — the last one has only digits and still overflows.
 */
const I32_MIN = -2147483648;
const I32_MAX = 2147483647;

function parseI32(raw) {
  if (!/^[+-]?\d+$/.test(raw)) return null;
  const n = Number(raw);
  if (!Number.isInteger(n) || n < I32_MIN || n > I32_MAX) return null;
  return n;
}

/**
 * Axum's two rejection messages. They differ by EXTRACTOR, not by position: `Path<i32>` names the
 * value; `Path<(i32, i32)>` names the index as well. Transcribed from real 400 responses.
 */
function invalidPath(raw, index = null) {
  const message =
    index === null
      ? `Invalid URL: Cannot parse \`${raw}\` to a \`i32\``
      : `Invalid URL: Cannot parse value at index ${index} with value \`${raw}\` to a \`i32\``;
  return new RejectionError(message, 400);
}

/** GET /v1/quran/surahs — quran.rs:21 */
export async function listSurahs(_req, reply, ctx) {
  const rows = await ctx.db.sql`
    SELECT a.surah_number,
           COUNT(*) AS ayah_count,
           s.english_name,
           s.name_arabic,
           s.english_translation,
           s.revelation_type
    FROM canonical_ayahs a
    LEFT JOIN canonical_surahs s ON s.surah_number = a.surah_number
    GROUP BY a.surah_number, s.english_name, s.name_arabic, s.english_translation, s.revelation_type
    ORDER BY a.surah_number`;

  // `SurahInfo` is a `#[derive(Serialize)]` STRUCT, so serde emits its fields in DECLARATION order
  // — not alphabetically like the `json!` responses below. Two different orders from the same file.
  // The key order here is that declaration order and is wire contract.
  const surahs = rows.map((r) => {
    const number = r.surah_number ?? 0;
    return {
      surahNumber: number,
      // COUNT(*) is int8; the driver hands it back as a string. Rust reads it as i64 and casts.
      ayahCount: Number(r.ayah_count ?? 0),
      // `.filter(|n| !n.is_empty())` — an EMPTY english_name falls back to the placeholder too, not
      // just a NULL one. The three fields below have no such filter: empty stays empty.
      name: r.english_name ? r.english_name : `Surah ${number}`,
      arabicName: r.name_arabic ?? "",
      translation: r.english_translation ?? "",
      revelationType: r.revelation_type ?? "",
    };
  });

  return reply.send(surahs);
}

/** GET /v1/quran/surahs/{surah_number} — quran.rs:80 */
export async function getSurah(req, reply, ctx) {
  const surahNumber = parseI32(req.params.surah_number);
  if (surahNumber === null) throw invalidPath(req.params.surah_number);

  const rows = await ctx.db.sql`
    SELECT id, surah_number, ayah_number, text_uthmani, source_checksum
    FROM canonical_ayahs WHERE surah_number = ${surahNumber} ORDER BY ayah_number`;

  if (rows.length === 0) throw NotFound();

  // `json!` keys, alphabetical: serde_json without `preserve_order` is BTreeMap-backed.
  return reply.send({
    ayahs: rows.map((r) => ({
      ayahNumber: r.ayah_number ?? 0,
      id: r.id ?? "",
      sourceChecksum: r.source_checksum ?? "",
      surahNumber: r.surah_number ?? 0,
      text: r.text_uthmani ?? "",
    })),
    surahNumber,
  });
}

/** GET /v1/quran/ayahs/{surah_number}/{ayah_number} — quran.rs:116 */
export async function getAyah(req, reply, ctx) {
  // `Path<(i32, i32)>`: the tuple extractor reports the INDEX of the segment that failed, and it
  // fails on the first bad one, left to right.
  const surahNumber = parseI32(req.params.surah_number);
  if (surahNumber === null) throw invalidPath(req.params.surah_number, 0);
  const ayahNumber = parseI32(req.params.ayah_number);
  if (ayahNumber === null) throw invalidPath(req.params.ayah_number, 1);

  const [row] = await ctx.db.sql`
    SELECT id, surah_number, ayah_number, text_uthmani, source_checksum
    FROM canonical_ayahs
    WHERE surah_number = ${surahNumber} AND ayah_number = ${ayahNumber}`;
  if (!row) throw NotFound();

  const words = await ctx.db.sql`
    SELECT id, word_index, text_uthmani, source_checksum
    FROM canonical_words WHERE ayah_id = ${row.id} ORDER BY word_index`;

  return reply.send({
    ayahNumber: row.ayah_number ?? 0,
    id: row.id,
    sourceChecksum: row.source_checksum ?? "",
    surahNumber: row.surah_number ?? 0,
    text: row.text_uthmani ?? "",
    words: words.map((w) => ({
      id: w.id ?? "",
      sourceChecksum: w.source_checksum ?? "",
      text: w.text_uthmani ?? "",
      wordIndex: w.word_index ?? 0,
    })),
  });
}
